kernel/signal.c

```diff
@@ -0,0 +1,552 @@
+	switch (how) {
+	case K_SIG_BLOCK:
+		k_sig_orset(&tid->base.sig_mask, set);
+		break;
+	case K_SIG_SETMASK:
+		tid->base.sig_mask = *set;
+		break;
+	case K_SIG_UNBLOCK:
+		k_sig_unset(&tid->base.sig_mask, set);
+		break;
+	}
+
+	return ret;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_sig_mask(int how, const struct k_sig_set *set, struct k_sig_set *oset)
+{
+	if (set != NULL) {
+		K_OOPS(K_SYSCALL_MEMORY_READ(set, sizeof(*set)));
+	}
+
+	if (oset != NULL) {
+		K_OOPS(K_SYSCALL_MEMORY_WRITE(oset, sizeof(*oset)));
+	}
+
+	return z_impl_k_sig_mask(how, set, oset);
+}
+#include <syscalls/k_sig_mask_mrsh.c>
+#endif
+
+int z_impl_k_sig_pending(struct k_sig_set *set)
+{
+	K_SPINLOCK(k_sig_lock) {
+		k_sig_pending_unlocked(set);
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_USERSPACE
+int z_vrfy_k_sig_pending(struct k_sig_set *set)
+{
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(set, sizeof(*set)));
+	return z_impl_k_sig_pending(set);
+}
+#include <syscalls/k_sig_pending_mrsh.c>
+#endif
+
+int k_sig_addset(struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	set->sig[i] |= BIT(bit);
+
+	return 0;
+}
+
+int k_sig_delset(struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	set->sig[i] &= ~BIT(bit);
+
+	return 0;
+}
+
+int k_sig_ismember(const struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	return (set->sig[i] & BIT(bit)) != 0;
+}
+
+int k_sig_fillset(struct k_sig_set *set)
+{
+	if (set == NULL) {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
+		set->sig[i] = -1;
+	}
+
+	return 0;
+}
+
+int k_sig_emptyset(struct k_sig_set *set)
+{
+	if (set == NULL) {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
+		set->sig[i] = 0;
+	}
+
+	return 0;
+}
+
+static void k_sig_orset(struct k_sig_set *result, const struct k_sig_set *set2)
+{
+	ARRAY_FOR_EACH(result->sig, i) {
+		result->sig[i] |= set2->sig[i];
+	}
+}
+
+static void k_sig_unset(struct k_sig_set *result, const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(result->sig, i) {
+		result->sig[i] &= ~set->sig[i];
+	}
+}
+
+static bool k_sig_setisempty(const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(set->sig, i) {
+		if (POPCOUNT(set->sig[i]) > 0) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static int k_sig_min_rt_in_set(const struct k_sig_set *set)
+{
+	for (int i = 0; i < K_SIG_NUM_RT; ++i) {
+		if (k_sig_ismember(set, (K_SIG_RTMIN + i)) > 0) {
+			return (K_SIG_RTMIN + i);
+		}
+	}
+	return 0;
+}
+
+static const char *k_strsigset_r(const struct k_sig_set *set, char *buf, size_t len)
+{
+	if ((set == NULL) || (buf == NULL) || (len < (K_STRSIGSET_CHARS + 1))) {
```

Comment:
As far as I could see the only place that this function is called ensures that `buf` is not `NULL` and `len` is correct.

Reply:
Fixed! (removed the function in favour of `LOG_HEXDUMP_DBG()`)